stellargraph/core/graph.py

```diff
@@ -832,6 +833,13 @@ def _unique_type_triples(self, *, return_counts, selector=slice(None)):
 
         return zip(*unique_ets)
 
+    def _edge_metrics_by_type_triple(self, *, selector=slice(None), metrics):
+        src_ty, rel_ty, tgt_ty = self._edge_type_triples(selector)
+        df = pd.DataFrame(
+            self._edges.weights, index=[src_ty, rel_ty, tgt_ty], columns=["weight"]
```

Comment:
This might be slightly clearer just using columns:

```suggestion
            {"src_ty": src_ty, "rel_ty": rel_ty, "tgt_ty": rel_ty, "weight": self._edges.weights}
```

and then `df.groupby(["src_ty", "rel_ty", "tgt_ty"])` on the next line.

It also turns out to be slightly faster because constructing the MultiIndex seems to be quite expensive. However, the difference is small enough that we should be focusing on the clarity, so I'm happy to go with whatever you prefer.

The following tables show using 4 columns (`columns`) vs the weight column + multiindex (`index`), for various numbers of edges (`n`), for creating the DataFrame (`creation`), running the groupby on an existing data frame (`groupby`) and the sum of the two (`total`). Times in milliseconds averaged across 100 runs.

1 node type, 1 edge type:
<table>
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">creation</th>
      <th colspan="2" halign="left">groupby</th>
      <th colspan="2" halign="left">total</th>
    </tr>
    <tr>
      <th></th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>0.68</td>
      <td>1.56</td>
      <td>1.58</td>
      <td>1.24</td>
      <td>2.26</td>
      <td>2.79</td>
    </tr>
    <tr>
      <th>100</th>
      <td>0.53</td>
      <td>1.27</td>
      <td>1.69</td>
      <td>1.27</td>
      <td>2.22</td>
      <td>2.54</td>
    </tr>
    <tr>
      <th>1000</th>
      <td>0.54</td>
      <td>1.32</td>
      <td>1.43</td>
      <td>1.18</td>
      <td>1.97</td>
      <td>2.49</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>0.53</td>
      <td>1.32</td>
      <td>1.73</td>
      <td>1.39</td>
      <td>2.27</td>
      <td>2.70</td>
    </tr>
    <tr>
      <th>100000</th>
      <td>2.62</td>
      <td>2.82</td>
      <td>4.25</td>
      <td>6.83</td>
      <td>6.87</td>
      <td>9.65</td>
    </tr>
    <tr>
      <th>1000000</th>
      <td>5.27</td>
      <td>31.72</td>
      <td>48.55</td>
      <td>29.56</td>
      <td>53.83</td>
      <td>61.28</td>
    </tr>
  </tbody>
</table>

1 node type, 10 edge types:

<table>
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">creation</th>
      <th colspan="2" halign="left">groupby</th>
      <th colspan="2" halign="left">total</th>
    </tr>
    <tr>
      <th></th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>0.64</td>
      <td>1.32</td>
      <td>1.76</td>
      <td>1.21</td>
      <td>2.40</td>
      <td>2.53</td>
    </tr>
    <tr>
      <th>100</th>
      <td>0.57</td>
      <td>1.22</td>
      <td>1.49</td>
      <td>1.23</td>
      <td>2.05</td>
      <td>2.46</td>
    </tr>
    <tr>
      <th>1000</th>
      <td>0.53</td>
      <td>1.22</td>
      <td>1.46</td>
      <td>1.27</td>
      <td>1.99</td>
      <td>2.49</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>0.68</td>
      <td>1.41</td>
      <td>1.74</td>
      <td>1.30</td>
      <td>2.42</td>
      <td>2.71</td>
    </tr>
    <tr>
      <th>100000</th>
      <td>1.24</td>
      <td>4.01</td>
      <td>4.74</td>
      <td>6.25</td>
      <td>5.97</td>
      <td>10.26</td>
    </tr>
    <tr>
      <th>1000000</th>
      <td>5.37</td>
      <td>32.43</td>
      <td>51.92</td>
      <td>31.78</td>
      <td>57.29</td>
      <td>64.21</td>
    </tr>
  </tbody>
</table>


10 node types, 10 edge types:

<table>
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">creation</th>
      <th colspan="2" halign="left">groupby</th>
      <th colspan="2" halign="left">total</th>
    </tr>
    <tr>
      <th></th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
      <th>columns</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>0.62</td>
      <td>1.40</td>
      <td>1.46</td>
      <td>1.27</td>
      <td>2.08</td>
      <td>2.67</td>
    </tr>
    <tr>
      <th>100</th>
      <td>0.58</td>
      <td>1.21</td>
      <td>1.57</td>
      <td>1.21</td>
      <td>2.15</td>
      <td>2.42</td>
    </tr>
    <tr>
      <th>1000</th>
      <td>0.53</td>
      <td>1.21</td>
      <td>1.59</td>
      <td>1.20</td>
      <td>2.11</td>
      <td>2.41</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>0.63</td>
      <td>1.46</td>
      <td>2.06</td>
      <td>1.43</td>
      <td>2.68</td>
      <td>2.89</td>
    </tr>
    <tr>
      <th>100000</th>
      <td>3.13</td>
      <td>6.28</td>
      <td>4.82</td>
      <td>3.96</td>
      <td>7.95</td>
      <td>10.24</td>
    </tr>
    <tr>
      <th>1000000</th>
      <td>4.50</td>
      <td>30.51</td>
      <td>52.23</td>
      <td>27.22</td>
      <td>56.73</td>
      <td>57.72</td>
    </tr>
  </tbody>
</table>

<details><summary>Code</summary>

```python
import pandas as pd

import numpy as np
import timeit

def time(g, rep=100):
    return timeit.timeit(g, number=rep) / rep

def f(N, etypes, ntypes, rep=100):
    s = np.random.randint(ntypes, size=N)
    e = np.random.randint(etypes, size=N)
    t = np.random.randint(ntypes, size=N)
    w = np.random.rand(N)

    col = lambda: pd.DataFrame({"stype": s, "etype": e, "ttype": t, "weight": w})

    col_create = time(col)
    df = col()
    col_group = time(lambda: df.groupby(["stype", "etype", "ttype"])["weight"].mean())
    
    ind = lambda: pd.DataFrame(w, index=[s, e, t], columns=["weight"])
    ind_create = time(ind)
    df = ind()
    ind_group = time(lambda: pd.DataFrame(w, index=[s, e, t], columns=["weight"]))
    return (col_create, ind_create, col_group, ind_group, col_create + col_group, ind_create + ind_group)

ns = [10, 100, 1000, 10000, 100000, 1000000]
columns = pd.MultiIndex.from_product([["creation", "groupby", "total"], ["columns", "index"]])

times_1_1 = pd.DataFrame([f(n, 1, 1) for n in ns], index=ns, columns=columns)
times_1_10 = pd.DataFrame([f(n, 1, 10) for n in ns], index=ns, columns=columns)
times_10_10 = pd.DataFrame([f(n, 10, 10) for n in ns], index=ns, columns=columns)

for x in [times_1_1, times_1_10, times_10_10]:
    print((x * 1000).round(2).to_html().replace(' border="1" class="dataframe"', ""))
```

</details>

Reply:
Nice! I'm happy to go with columns, I think it looks clearer too avoiding the `level=[0,1,2]`